teachopencadd/talktorials/T018_automated_cadd_pipeline/utils/binding_site_detection.py

```diff
@@ -0,0 +1,244 @@
+        )
+        # calculate the geometric center of the molecule,
+        # which represents the center of the rectangular box,
+        # as well as the length of the molecule in each dimension,
+        # which corresponds to the edge lengths of the rectangular box.
+        # Also, we will add a 5 Å buffer in each dimension to allow the
+        # correct placements of ligands that are bigger than the co-crystallized ligands
+        # or bind in a different fashion.
+        pocket_center = (
+            ligand_object.positions.max(axis=0) + ligand_object.positions.min(axis=0)
+        ) / 2
+        pocket_size = ligand_object.positions.max(axis=0) - ligand_object.positions.min(axis=0) + 5
+        pocket_coordinates = {
+            "center": pocket_center.tolist(),
+            "size": [pocket_size.tolist()],
+        }
+        Protein.binding_site_coordinates = pocket_coordinates
+        return
+
+    def compute_by_detection(self, Protein, BindingSiteSpecs, binding_site_output_path):
+        """
+        TODO Add docstring
+        """
+        # derive the relevant function name from detection method
+        detection_method_name = f"detect_by_{BindingSiteSpecs.detection_method.name.lower()}"
+        # get the function from its name
+        detection_method = getattr(self, detection_method_name)
+        # call the function
+        detection_method(Protein, BindingSiteSpecs, binding_site_output_path)
+
+    def detect_by_dogsitescorer(self, Protein, BindingSiteSpecs, binding_site_output_path):
+        """
+        TODO Add docstring
+        """
+        if hasattr(Protein, "pdb_code"):
+            self.dogsitescorer_pdb_id = Protein.pdb_code
+        elif hasattr(Protein, "pdb_filepath"):
+            self.dogsitescorer_pdb_id = dogsitescorer.upload_pdb_file(Protein.pdb_filepath)
+        else:
+            raise AttributeError(f"Protein has neither `pdb_code` nor `pdb_filepath attributes.")
+
+        # try to get the chain_id for binding site detection if it's available in input data
+        if BindingSiteSpecs.protein_chain_id != "":
+            if hasattr(Protein, "chains") and BindingSiteSpecs.protein_chain_id in Protein.chains:
+                self.dogsitescorer_chain_id = BindingSiteSpecs.protein_chain_id
+            else:
+                raise ValueError(
+                    f"The input protein chain-ID ({BindingSiteSpecs.protein_chain_id}) "
+                    f"does not exist in the input protein. Existing chains are: {Protein.chains}"
+                )
+        else:
+            # if chain_id is not in input data, try to set it to the first chain found in PDB file
+            try:
+                self.dogsitescorer_chain_id = Protein.chains[0]
+            # if no chain is found in PDB file either, leave the chain_id empty
+            except:
+                # FIXME specify exception
+                self.dogsitescorer_chain_id = ""
```

Comment:
@Armin-Ariamajd, what kind of exception do you expect here? `KeyError` because there are no chains or `AttributeError` because we do not even have the attribute `chain`?

Reply:
It may be that the protein does not have any chain information in the PDB file, which depending on the context of the PDB file will result in either `Protein.chains` to be an empty list (in which case `Protein.chains[0]` will throw an `IndexError: list index out of range`), or it may be that `Protein.chains` does not get created, in which case we will get an `AttributeError`. 
We don't want to raise an exception here though, because the dogsitescorer will also work without providing a chain-ID (and in case of a protein without any chain information, there is no chain-ID to be provided anyway), hence the:
```python
except:
   self.dogsitescorer_chain_id = ""
```
I don't see a case where this can cause a problem in running the code.